# 0.1 Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 0.2 Importing Dataframes

In [2]:
# Set filepath for IC Basket project
path = r'/Users/joefeuille/CF Analysis/Instacart Basket Analysis'

In [3]:
# Set Orders set as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
# Set Products set as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords.shape

(3421083, 8)

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_time_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# Mixed Data Types

In [6]:
# Create a data frame
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [10]:
# Reassigning "mix" column as a string variable
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [11]:
# Searching Products df for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create new subset containing null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.shape

(49693, 5)

In [15]:
# Creating new df without null values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [17]:
# Creating a new subset containing only duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
# Checking for full duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
# Create new df with no dups
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
# Renaming checked set as something that makes sense
df_prods_checked = df_prods_clean_no_dups

In [22]:
df_prods_checked.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# TASK 4.5

## Q.1: If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

In [23]:
### Done. Please see above.

## Q.2: Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. 

In [24]:
# Checking original df first.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [25]:
# Checking new df for comparison.
df_prods_checked.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

### Most of the values seem consistent, except that I suspect at least one outlier in the prices column. There's no way a grocery store carries something that's just shy of $100k.
### It might be worth checking the number of departments for any gaps, and the number of aisles too. A grocery store could have 134 aisles, but it's worth checking.

In [26]:
# Checking for prices column first.
df_prods_checked.sort_values('prices', ascending=False)


product_id                      product_name  aisle_id  department_id  \
33666       33664             2 % Reduced Fat  Milk        84             16   
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
19392       19391         Turkey Breast Tenderloins        49             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   
40490       40486                   Chicken Tenders        49             12   
...           ...                               ...       ...            ...   
11307       11307                  Goat Cheese Logs         2             16   
32805       32803                 Regular Deodorant        80             11   
46627       46623                   Pie Pans, Large        10             17   
37041       37037                       Redseedless        24              4   
21464       21463      Vanilla Bean Light Ice Cream        37              1   

        prices  
33666  99999.0  
21554  14900.0  
19392     25.0  
25580     25.0  
40490     25.0  
...        ...  
11307      1.0  
32805      1.0  
46627      1.0  
37041      1.0  
21464      1.0  

[49672 rows x 5 columns]

In [27]:
# Yep. Nobody's gonna pay that much for milk or cottage cheese.

In [28]:
# Checking for aisle_id column.
df_prods_checked.sort_values('aisle_id', ascending=False)

product_id                                       product_name  \
45285       45281                                            Marsala   
12825       12825                                            Sangria   
6105         6105                                        Junmai Sake   
22020       22019                                   Six Grapes Porto   
2251         2251                                          Ruby Port   
...           ...                                                ...   
34239       34237             Organic Add Me Kale Blend Bagged Salad   
4112         4112  Chopped Salad Bowl Italian Salad with Salami &...   
41120       41116                                Spinach Dijon Salad   
22854       22853                                  Pesto Pasta Salad   
15767       15767                            Blackened Chicken Salad   

       aisle_id  department_id  prices  
45285       134              5    10.8  
12825       134              5     8.0  
6105        134              5     5.6  
22020       134              5    10.8  
2251        134              5     8.5  
...         ...            ...     ...  
34239         1             20     6.9  
4112          1             20     9.1  
41120         1             20    10.8  
22854         1             20     6.8  
15767         1             20    11.1  

[49672 rows x 5 columns]

In [29]:
df_prods_aisles = df_prods_checked.sort_values('aisle_id', ascending=False)

In [30]:
df_prods_aisles.head(30)

product_id                                product_name  aisle_id  \
45285       45281                                     Marsala       134   
12825       12825                                     Sangria       134   
6105         6105                                 Junmai Sake       134   
22020       22019                            Six Grapes Porto       134   
2251         2251                                   Ruby Port       134   
44960       44956                                 Red Moscato       134   
44875       44871                       Moscato Spumante Wine       134   
30584       30582                           Natural Champagne       134   
4285         4285          Sangria, Tradicional, Fresh Citrus       134   
27908       27906                          Imperial Champagne       134   
2216         2216                        California Champagne       134   
39615       39611                      Blackberry Merlot Wine       134   
34940       34938                              Côtes-Du-Rhône       134   
2989         2989                               Sweet Marsala       134   
30566       30564                                   Port Wine       134   
12586       12586                          Amontillado Sherry       134   
29612       29610                           Cordon Negro Brut       134   
2590         2590                        Grand Brut Champagne       134   
1808         1808                                   Champagne       134   
19541       19540                              Sparkling Brut       134   
40369       40365                      American Concord Grape       134   
24363       24362                         Brut Sparkling Wine       134   
40676       40672                   Brut California Champagne       134   
19804       19803                  Sparking Brut Classic Wine       134   
7203         7203                                  Tawny Port       134   
46729       46725  Champagne, Extra Dry, Sparkling California       134   
39043       39039                           Organic Nama Sake       134   
46384       46380                         Champagne Brut Rose       134   
20089       20088                                Brut Premier       134   
2022         2022              Brut Cuvee Sparkling Champagne       134   

       department_id  prices  
45285              5    10.8  
12825              5     8.0  
6105               5     5.6  
22020              5    10.8  
2251               5     8.5  
44960              5     8.4  
44875              5     5.2  
30584              5     2.2  
4285               5     4.0  
27908              5    13.6  
2216               5     1.8  
39615              5     7.2  
34940              5     3.7  
2989               5     2.5  
30566              5     2.1  
12586              5     8.1  
29612              5    12.4  
2590               5     3.1  
1808               5     4.4  
19541              5    14.6  
40369              5    13.3  
24363              5     7.6  
40676              5    12.6  
19804              5    13.6  
7203               5     5.3  
46729              5     8.1  
39043              5     2.9  
46384              5     3.5  
20089              5    10.9  
2022               5     1.4

### I suspect the aisles column may actually be correct. But it is definitely worth revisiting those to pricing outliers.

# Q.3: Check for mixed-type data in your df_ords dataframe.

In [31]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_time_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [32]:
# Check for mixed types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

#### Mix check returned nothing. Therefore, there were no columns with mixed types.

# Q.5: Run a check for missing values in your df_ords dataframe. 

##    In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [33]:
# Searching Orders df for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_time_of_day              0
days_since_prior_order    206209
dtype: int64

### Days Since Prior Order has 206,209 missing values.

In [34]:
# Create new subset containing null values
df_nulls = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [35]:
df_nulls.head(20)

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0             0   2539329        1    prior             1                   2   
11           11   2168274        2    prior             1                   2   
26           26   1374495        3    prior             1                   1   
39           39   3343014        4    prior             1                   6   
45           45   2717275        5    prior             1                   3   
50           50   2086598        6    prior             1                   5   
54           54   2565571        7    prior             1                   3   
75           75    600894        8    prior             1                   6   
79           79    280530        9    prior             1                   1   
83           83   1224907       10    prior             1                   2   
89           89   2411567       11    prior             1                   0   
97           97   1483446       12    prior             1                   1   
103         103   2618231       13    prior             1                   6   
116         116   2906263       14    prior             1                   2   
130         130   2514324       15    prior             1                   5   
153         153   1273676       16    prior             1                   2   
160         160   1737705       17    prior             1                   2   
201         201   2780889       18    prior             1                   6   
208         208   1321974       19    prior             1                   6   
218         218    947722       20    prior             1                   1   

     order_time_of_day  days_since_prior_order  
0                    8                     NaN  
11                  11                     NaN  
26                  14                     NaN  
39                  11                     NaN  
45                  12                     NaN  
50                  18                     NaN  
54                   9                     NaN  
75                   0                     NaN  
79                  17                     NaN  
83                  14                     NaN  
89                  11                     NaN  
97                  12                     NaN  
103                 12                     NaN  
116                 18                     NaN  
130                 13                     NaN  
153                  8                     NaN  
160                 13                     NaN  
201                 18                     NaN  
208                 13                     NaN  
218                 10                     NaN

In [36]:
df_nulls.tail(20)

Unnamed: 0  order_id  user_id eval_set  order_number  \
3420678     3420678   1955107   206190    prior             1   
3420687     3420687    936117   206191    prior             1   
3420693     3420693   3242108   206192    prior             1   
3420708     3420708    313609   206193    prior             1   
3420750     3420750   3395125   206194    prior             1   
3420762     3420762    926652   206195    prior             1   
3420782     3420782    825502   206196    prior             1   
3420787     3420787   3325493   206197    prior             1   
3420811     3420811   2154947   206198    prior             1   
3420819     3420819   3302466   206199    prior             1   
3420839     3420839   2508134   206200    prior             1   
3420863     3420863     28210   206201    prior             1   
3420896     3420896    799173   206202    prior             1   
3420919     3420919   1991850   206203    prior             1   
3420925     3420925   1438269   206204    prior             1   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_time_of_day  days_since_prior_order  
3420678                   6                  9                     NaN  
3420687                   6                 15                     NaN  
3420693                   6                  6                     NaN  
3420708                   5                 15                     NaN  
3420750                   1                  9                     NaN  
3420762                   3                 16                     NaN  
3420782                   6                 14                     NaN  
3420787                   2                 19                     NaN  
3420811                   0                 16                     NaN  
3420819                   2                 11                     NaN  
3420839                   5                 19                     NaN  
3420863                   6                 14                     NaN  
3420896                   2                 12                     NaN  
3420919                   1                 13                     NaN  
3420925                   1                 11                     NaN  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN

### None of thise user_id numbers are repeating. This might suggest that the nulls values are associated with each user's first order.

# Q.6: Address the missing values using an appropriate method.

In [37]:
df_ords.shape

(3421083, 8)

In [38]:
df_nulls.shape

(206209, 8)

In [40]:
#Adding a true/false column for First Time Orders
df_ords['first_time_ordering'] = np.where(df_ords['days_since_prior_order'].isnull(), True, False)

In [41]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_time_of_day  days_since_prior_order  first_time_ordering  
0                  8                     NaN                 True  
1                  7                    15.0                False  
2                 12                    21.0                False  
3                  7                    29.0                False  
4                 15                    28.0                False

### Added a new column for "First Time Ordering" to keep the sales data for further analysis.

# Q.7: Run a check for duplicate values in your df_ords data. 

In [43]:
# Checking non-null order df for duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [44]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_time_of_day, days_since_prior_order, first_time_ordering]
Index: []

### No dups found in ords_clean

# Q.9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

### df_prods was exported earlier. Pleasee see above.

In [45]:
df_ords_clean = df_ords

In [46]:
# Exporting ords_clean as Orders Checked
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))